Second wave of API changes

Meziu opened the pull request:
Related to #102 

This is a second wave of breaking API changes based on the official [guidelines](https://rust-lang.github.io/api-guidelines/checklist.html) (I can hear you cry from here).

Most notable changes:
- All service `init` functions have been renamed to `new` (it's standard for any constructor).
- Types derive as many traits as possible (each was thought out for the type's specific purpose, of course)
- Small renames
- `Gfx::swap_buffers` and `Gfx::flush_buffers` have been moved to the `Screen` trait. Originally, those functions modified a global state. Now each `Screen` can flush its own buffers. Also, `Console` handles flushing and swapping by itself, so most examples don't even mention those functions anymore.

This is one of the last PRs in preparation to the docs PR.

Meziu committed:
Update for linker-fix-3ds rename

Meziu committed:
fmt

Meziu committed:
Renamed init functions to new

Meziu committed:
Renamed init functions to new

Meziu committed:
Renamed init functions to new

Meziu committed:
Derive traits and Screen changes

Meziu committed:
Derive traits and Screen changes

Meziu committed:
Use usize in RawFrameBuffer

Meziu committed:
Use usize in RawFrameBuffer

Meziu committed:
Renamed SystemModel enum variants

Meziu committed:
Renamed SystemModel enum variants

Meziu committed:
Fmt and clippy

Meziu committed:
Fmt and clippy

Meziu committed:
Undo double flush in example

Meziu committed:
Undo double flush in example

Meziu committed:
Compilation error when romfs can't be found

Meziu committed:
fmt

Meziu committed:
Compilation error only if feature is set

Meziu committed:
fmt

Meziu committed:
Explicit DPAD nomenclature

Meziu committed:
Explicit DPAD nomenclature

ian-h-chamberlain commented on ctru-rs/src/services/gfx.rs:
```diff
@@ -55,6 +55,28 @@ pub trait Screen: private::Sealed {
+    ///
+    /// This should be used even if double buffering is disabled.
+    fn swap_buffers(&mut self) {
+        unsafe { ctru_sys::gfxScreenSwapBuffers(self.side().into(), true) };
```
ah nice, I was wondering how these would work since the `gfxFlushBudders` etc were global, I didn't realize there are APIs for doing it per screen.

Should we set the `hasStereo` parameter to `false` here by default, and set to `true` only if the screen in question is the "combined" top screen?

Meziu replied:
That would require quite a bit of work on the `TopScreen3D` struct, since it doesn't implement `Screen`. Since you built it, do you have any idea how we could do so?

ian-h-chamberlain replied:
See my other comment, I think we might be able to do it with a separate trait instead of having these methods on `Screen`, but it might take a bit of experimenting, so maybe better off in a separate PR.

ian-h-chamberlain commented on ctru-rs/src/services/gfx.rs:
```diff
@@ -96,14 +118,14 @@ pub struct RawFrameBuffer<'screen> {
+    pub width: usize,
     /// The height of the framebuffer in pixels.
-    pub height: u16,
+    pub height: usize,
```
IIRC these were `u16` since the maximum value is much smaller than `usize::MAX` but it's probably fine to have like this as long as there is some check for max values somewhere

Meziu replied:
Yes, it's always smaller that `u16::MAX`. However, since this is a "read-only" member (there is no function that asks for these sizes, right?) there is no problem using it this way. The user will convert it back to `usize` 99% of times since that needed to work with pointers. It's just easier to have it like this.

ian-h-chamberlain commented on ctru-rs/src/services/gfx.rs:
```diff
@@ -55,6 +55,28 @@ pub trait Screen: private::Sealed {
         unsafe { ctru_sys::gfxSetDoubleBuffering(self.as_raw(), enabled) }
     }
 
+    /// Swaps the video buffers.
```
```suggestion
    /// Flushes the video buffer for this screen.
```

Meziu replied:
Copying and pasting is always a bad habit... thanks for noticing (though I will look into all docs later on) 👍

ian-h-chamberlain commented on ctru-rs/src/services/cfgu.rs:
```diff
@@ -33,15 +33,15 @@ pub enum Language {
-    Model2DS = ctru_sys::CFG_MODEL_2DS,
-    ModelNew3DSXL = ctru_sys::CFG_MODEL_N3DSXL,
-    ModelNew2DSXL = ctru_sys::CFG_MODEL_N2DSXL,
+    N3DS = ctru_sys::CFG_MODEL_3DS,
```
This might be slightly confusing since sometimes I have seen `N3DS` used to mean `New 3DS`, but I suppose the fact that there is a separate `NewN3DS` should make that clear. Don't have a better suggestion really

Meziu replied:
I went with Old3DS. It's not official but people recognize it and it's much less confusing.

ian-h-chamberlain commented on ctru-rs/src/applets/swkbd.rs:
```diff
@@ -7,6 +7,7 @@ use std::iter::once;
 use std::str;
 
 /// An instance of the software keyboard.
+#[derive(Clone)]
```
does cloning the inner Box here end up cloning all the global state too? I would guess normally we don't want `Clone` for these "global-ish" kind of structs...

Meziu replied:
Today in "Wacky facts about libctru" we will learn: Software Keyboards!

Did you know? The "SwkbdState" struct only holds config data, since the Software Keyboard is "spawned" on request, and is not a global service.

Next episode, we will find out why internal atomic counters aren't accessible via the API (small spoiler: there is no reason! 🥳 )

ian-h-chamberlain commented on ctru-rs/src/applets/mii_selector.rs:
```diff
@@ -8,14 +8,14 @@ use std::ffi::CString;
-#[derive(Debug, Clone)]
-pub enum MiiConfigIndex {
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub enum MiiIndex {
```
this could probably even just be called `mi_selector::Index`, same for other structs in this module.

Or actually, we could call the module `mii` and then you'd have `mii::Type`, `mii::Selector`, etc.

Meziu replied:
The mii module already exists, lol. I can rename it to settle for a better `module::struct` nomenclature.

ian-h-chamberlain replied:
ahh I didn't realize that, maybe it should just be `mii::selector::Index` etc then

Techie-Pi replied:
Yeah, `mii:selector` seems like a good option tbh

Meziu replied:
That would move the module out of the `applets` folder. Do you think that would be a good idea? There are many more applets left to be implemented.

Meziu replied:
I simply renamed it to `Index` for now. Tell me if you think more changes are needed.

ian-h-chamberlain commented on ctru-rs/examples/hello-world.rs:
```diff
@@ -29,9 +29,6 @@ fn main() {
         }
-        // Flush and swap framebuffers
-        gfx.flush_buffers();
-        gfx.swap_buffers();
```
you mentioned that the console handles flushing the screen automatically... does it do that whenever the stdout is flushed (e.g. on `\n` or explicit `flush()` calls)? 

I'd imagine that for most applications (non-console) the user would still want to flush + swap buffers every frame instead of letting underlying helpers do it at random times.

Edit: also, from looking at https://github.com/devkitPro/libctru/blob/master/libctru/source/console.c it seems the buffers are flushed by default but not normally swapped, which I guess doesn't matter since `consoleInit` also sets double-buffering off...

Meziu replied:
We could specify that behaviour in a comment, since there is no need for manually flushing. Console does it on `print()`. I'll put this off for the docs PR. 

Edit: Overall this goes nicely with our "screen ownership" model: `Console` has it, `Console` manages it. If the user has a different screen (like in `graphics-bitmap`) they can manage that.

ian-h-chamberlain commented on ctru-rs/examples/gfx-3d-mode.rs:
```diff
@@ -61,9 +61,8 @@ fn main() {
-        gfx.flush_buffers();
-        gfx.swap_buffers();
+        left.flush_buffer();
+        left.swap_buffers();
```
this probably should be flushing whichever side is being drawn to, not just the left screen?

Meziu replied:
The issue is with the "stereo" parameter. Since the screen is in 3D mode, and the `TopRight` screen returns "top" as its raw value, the swap/flush would be done twice on the top screen, once when "Flushing left" and once when "flushing right" (it flushes both of them each time). I guess that means this function should live in `TopScreen3D`, but that doesn't implement `Screen`, so it's a bit of an issue.

ian-h-chamberlain replied:
Hmm, after rereading the documentation for the `libctru` functions I see what you mean. I guess in most cases people will want to swap both sides' buffers at the same time, so it's probably fine.

I guess we could awkwardly make the `TopLeft` and `TopRight` impls do nothing and force the user to call a special `flush_both` type method on the `TopScreen3D`, but I think we will run into mutability issues etc, plus having a method do nothing feels kinda wrong to me.

Another option is to define a totally new trait, maybe something like this:
```rs
pub trait Flush: private::Sealed {
    fn swap_buffers(&mut self);
    fn flush_buffer(&mut self);
}
```

I think we'd have to manually impl it for `TopScreen`, `BottomScreen` and `TopScreen3D` (which would have to flush both left + right sides). I'm not sure if there is a use case for flushing left/right separately, but if needed we could also split `Flush` and `Swap` to separate traits as well.

ian-h-chamberlain replied:
Update: I opened #118 to show what I am thinking of. We can discuss further on that PR

ian-h-chamberlain reviewed (commented):
Mostly looks good! I had some questions about the Gfx changes but I think everything else looks good.

Also you may want to merge `master` back into this branch to make the diff a bit simpler (without displaying the changes from #115 )

Meziu committed:
Update flush_buffer docs

Co-authored-by: Ian Chamberlain <ian-h-chamberlain@users.noreply.github.com>

Meziu committed:
Update flush_buffer docs

Co-authored-by: Ian Chamberlain <ian-h-chamberlain@users.noreply.github.com>

Meziu committed:
Update gfx.rs

Meziu committed:
Update gfx.rs

ian-h-chamberlain committed:
Initial pass of Swap/Flush traits for screens

ian-h-chamberlain committed:
Merge branch 'improve/api' into feat/flush-swap-screen-traits

ian-h-chamberlain committed:
Do some cleanup and fix swap trait impls

Oops! The argument to `gfxScreenSwapBuffers` is gfxScreen_t, not
gfx3dSide_t ! Unfortunately they are both alias to libc::c_uint so we
didn't get any help from the Rust compiler when passing the screen side
into this function.

ian-h-chamberlain committed:
On second thought, it seems better to reuse traits

Rename to flush_buffers, and just have TopScreen3D's impl do both. This
still allows individual flushing of left and right if needed but most
people would probably just use `TopScreen3D::flush_buffers`.

AzureMarker commented on ctru-rs/src/services/sslc.rs:
```diff
@@ -8,23 +8,12 @@ pub struct SslC(());
     }
-
-    /// Fill `buf` with `buf.len()` random bytes
-    pub fn generate_random_data(&self, buf: &mut [u8]) -> crate::Result<()> {
```
Is this supposed to get removed?

Meziu replied:
This is a duplicate of `Ps::generate_random_data`. The user should use that one or the `rand` crate, which works via `getrandom`.

ian-h-chamberlain committed:
Update doc comments and bitmap example

Clarify the behavior of `swap_buffers` for single buffering mode and
update bitmap example to prove that this is actually how it works.

ian-h-chamberlain committed:
Fix mutability issues after merge of master

Meziu committed:
Rename N3DS to Old3DS

Meziu committed:
Rename MiiIndex to Index

Meziu committed:
Fix doctest fmt

Meziu committed:
Fix doctest

Meziu committed:
HID Touch and Circlepad refactor + new example

Meziu committed:
Merge branch 'improve/api' into feat/flush-swap-screen-traits

Meziu committed:
Fix merge messages

Meziu committed:
Readd trait imports

Meziu committed:
Merge pull request #118 from rust3ds/feat/flush-swap-screen-traits

Create some traits for swapping/flushing screen buffers
